packages/adminrouter/extra/src/lib/cache.lua

```diff
@@ -238,23 +243,14 @@ local function fetch_and_store_marathon_apps(auth_token)
        end
 
        local ports = task["ports"] --task host port mapping by default
-       if is_ip_per_task(app) then
+       if is_container_network(app) then
+         -- override with ports from the container's portMappings
          ports = {}
-         if is_user_network(app) then
-            -- override with ports from the container's portMappings
-            local port_mappings = app["container"]["docker"]["portMappings"] or app["portDefinitions"] or {}
-            local port_attr = app["container"]["docker"]["portMappings"] and "containerPort" or "port"
-            for _, port_mapping in ipairs(port_mappings) do
-               table.insert(ports, port_mapping[port_attr])
-            end
-         else
-            --override with the discovery ports
-            local discovery_ports = app["ipAddress"]["discovery"]["ports"]
-            for _, discovery_port in ipairs(discovery_ports) do
-                table.insert(ports, discovery_port["number"])
-            end
-         end
-       end
+         local port_mappings = app["container"]["portMappings"] or {}
+          for _, port_mapping in ipairs(port_mappings) do
+             table.insert(ports, port_mapping["containerPort"])
```

Comment:
note:
- `containerPort` might be `0` (which means "marathon, please pick a 'free' container port for me")
- `hostPort` might also be `0` (which means "marathon, please pick a 'free' host port for me")

you'll probably need to carefully need to correlate such cases with the ports reported in the task status in order to determine the *actual* port to which a request should be sent. you'll want test cases for this

Reply:
for non-container-bridge mode, `hostPort` might not be defined at all (mode=container)

Reply:
I have added the routing logic for the cases you mentioned. Tested them manually, going to add unit tests shortly.

Reply:
@jdef I am not familiar with the marathon integration tests, do I understand it correctly that the following [test](https://github.com/dcos/dcos/blob/master/packages/dcos-integration-test/extra/test_networking.py#L407) is deploying a marathon app in "container" networking mode and trying to check if the container port exposed by the [test server](https://github.com/dcos/dcos/blob/master/packages/dcos-integration-test/extra/util/python_test_server.py) is reachable? Do I understand it correctly that Admin Router routing logic is **not** used there?

Reply:
i'm not super-familiar with these tests. maybe @kensipe or @zen-dog can comment here